modules/ardrivo/test/Arduino.cxx

```diff
@@ -0,0 +1,268 @@
+
+    auto loc_board_data = std::make_unique<BoardData>();
+    loc_board_data->silence_errors = false;
+    loc_board_data->start_time = std::chrono::steady_clock::now();
+    loc_board_data->pin_modes = std::vector<std::atomic_uint8_t>(3);
+    loc_board_data->digital_pin_values = std::vector<std::atomic_bool>(3);
+    loc_board_data->analog_pin_values = std::vector<std::atomic_uint16_t>(3);
+    loc_board_data->pin_frequency = std::vector<std::atomic_uint8_t>(3);
+    loc_board_data->interrupts_handlers = std::vector<std::pair<void (*)(), int>>(3);
+    auto loc_board_info = std::make_unique<BoardInfo>();
+    loc_board_info->pins_caps.resize(3);
+    ranges::for_each(loc_board_info->pins_caps, [](PinCapability& pin) {
+        pin.digital_in = pin.digital_out = true;
+        pin.analog_in = pin.analog_out = true;
+        pin.interrupt_able = pin.interrupt_able = true;
+    });
+
+    init(loc_board_data.release(), loc_board_info.release());
+}
+
+TEST_CASE("Read and write digital & analog & pinMode", "[pinMode], [digitalWrite], [digitalRead], [analogWrite], [analogRead]") {
+    init_fake();
+
+    SECTION("Set pinModes", "[pinMode]") 
+    {         
+        auto i = GENERATE(0, 1, 2);
+        board_data->pin_modes[i] = static_cast<uint8_t>(PinMode::OUTPUT);
+        REQUIRE(board_data->pin_modes[i] == 1U);    
+    }
+
+    SECTION("Set digital pins ") 
+    {
+        digitalWrite(1, LOW);
+        REQUIRE(digitalRead(1) == LOW);
+        digitalWrite(1, HIGH);
+        REQUIRE(digitalRead(1) == HIGH);
+
+        auto i = GENERATE(0, 1, 2);
+        digitalWrite(i, HIGH);
+        REQUIRE(digitalRead(i) == HIGH);     
+    }
+    SECTION("Set analog pins") {
+        analogWrite(1, 0);
+        REQUIRE(analogRead(1) == 0);
+        analogWrite(1, 500);
+        REQUIRE(analogRead(1) == 500);
+    }
+
+    SECTION("Set value of pins over vector size. digitalWrite and analogWrite") 
+    {
+        REQUIRE_THROWS_WITH(digitalWrite(32, HIGH), "digitalWrite(32, true): Pin does not exist");
+        REQUIRE_THROWS_WITH(analogWrite(32, 100), "analogWrite(32, 100): Pin does not exist");
+    }
+}
+
+TEST_CASE("Set tone frequency on pin", "[tone], [noTone]")
+{
+    init_fake();
+    auto i = GENERATE(0, 1, 2);
+    board_data->pin_frequency[i] = 30;
+    REQUIRE(board_data->pin_frequency[i] == 30);
+}
+
+
+// TEST_CASE("Pulse in");
+
+// TEST_CASE("Pulse long in ");
+
+// TEST_CASE("shift in");
+
+// TEST_CASE("shift out");
+
+TEST_CASE("Delay the program", "[delay]") 
+{
+    const auto start = std::chrono::steady_clock::now();
+    delay(1000);
+    const auto end = std::chrono::steady_clock::now();
+    const auto duration = std::chrono::duration_cast<std::chrono::seconds>(end - start);
+    //Added some margin
+
+    REQUIRE(duration > 0s);
+    REQUIRE(duration < 1.5s);
+}
+
+
+TEST_CASE("delay the thread in microseconds", "[delayMicroseconds]") 
+{
+    const auto start = std::chrono::steady_clock::now();
+    delay(1000);
```

Comment:
You're effectively testing `delay` but your test says you should be testing `delayMicroseconds`